hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ConsensusTimeHook.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.workflows;
+
+import static com.hedera.node.app.service.mono.ledger.accounts.staking.StakePeriodManager.DEFAULT_STAKING_PERIOD_MINS;
+import static com.hedera.node.app.service.mono.utils.Units.MINUTES_TO_MILLISECONDS;
+import static com.swirlds.common.stream.LinkedObjectStreamUtilities.getPeriod;
+import static java.time.ZoneOffset.UTC;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.node.app.service.token.impl.handlers.staking.EndOfStakingPeriodUpdater;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.config.data.StakingConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Class responsible for running any actions that need to happen at the end of
+ * transaction handling. The reason it's called a consensus time hook is because
+ * the actions are (possibly) triggered by checking the previous transaction's
+ * consensus time against the consensus time of the current transaction.
+ */
+@Singleton
+public class ConsensusTimeHook {
```

Comment:
As this is a staking period specific implementation, it might be good to split out an interface if we expect to create other implementations (e.g. `StakingPeriodTimeHook implements ConsensusTimeHook`).
If we do not expect to ever create other implementations, then the name perhaps should be StakingPeriodTimeHook.

Reply:
I hadn't thought about creating other implementations, just putting more into this class eventually. I could see it either way, and it feels cleaner that way when we add the missing stuff

Reply:
No need to block the PR, but it would be good to at least split interface and implementation if we are certain that more needs to be added later; so we can choose then how to split things.

Reply:
Done